Data/Root/test/RootDataIOTest.cpp

```diff
@@ -17,25 +24,183 @@
+                      SampleOut[i].ParticleList.front().fourMomentum().e());
+  }
+
+  std::remove(FileName);
+}
+
+BOOST_AUTO_TEST_CASE(ReadWildcards) {
+  ComPWA::Logging log("TRACE");
+
+  using namespace ComPWA::Data::Root;
+  auto Sample1 = generateSample(1.864, {0.5, 0.5, 0.5}, 10);
+  auto Sample2 = generateSample(1.864, {0.5, 0.5, 0.5}, 20);
+  auto Sample3 = generateSample(1.864, {0.5, 0.5, 0.5}, 30);
+  writeData(Sample1, "RootReaderTest-writeData1.root", "Correct");
+  writeData(Sample2, "RootReaderTest-writeData2.root", "WRONG");
+  writeData(Sample3, "RootReaderTest-writeData3.root", "Correct");
+  createRootTree("RootReaderTest-createRootTree1.root", "Correct", 15);
+  createRootTree("RootReaderTest-createRootTree2.root", "WRONG", 25);
+
+  auto SampleIn(readData("RootReaderTest-*.root", "Correct"));
+  BOOST_CHECK_EQUAL(SampleIn.size(), 55);
+
+  std::remove("RootReaderTest-writeData1.root");
+  std::remove("RootReaderTest-writeData2.root");
+  std::remove("RootReaderTest-writeData3.root");
+  std::remove("RootReaderTest-createRootTree1.root");
+  std::remove("RootReaderTest-createRootTree2.root");
+}
+
+BOOST_AUTO_TEST_CASE(WriteTwoTreesToSameFile) {
+  ComPWA::Logging log("TRACE");
+
+  using namespace ComPWA::Data::Root;
+  auto SampleOut1 = generateSample(1.864, {0.5, 0.5, 0.5}, 10);
+  auto SampleOut2 = generateSample(1.864, {0.5, 0.5, 0.5}, 20);
+
+  const char *OutputFileName = "RootReaderTest-sameFile.root";
+  writeData(SampleOut1, OutputFileName, "Tree1", false);
+  writeData(SampleOut2, OutputFileName, "Tree2", false);
+
+  TFile TheFile(OutputFileName);
+  if (TheFile.IsZombie())
+    throw std::runtime_error("Could not load file");
+  TList *ListOfKeys = TheFile.GetListOfKeys();
+  if (!ListOfKeys)
+    throw std::runtime_error("Empty list of keys");
+
+  BOOST_CHECK_EQUAL(ListOfKeys->GetEntries(), 2);
+  BOOST_CHECK_EQUAL(ListOfKeys->At(0)->GetName(), "Tree1");
+  BOOST_CHECK_EQUAL(ListOfKeys->At(1)->GetName(), "Tree2");
+
+  std::remove(OutputFileName);
+}
 
-  auto sample(ComPWA::Data::generatePhsp(200, gen, RandomGenerator));
+BOOST_AUTO_TEST_CASE(OpenWrongBranches) {
+  ComPWA::Logging log("TRACE");
 
-  ComPWA::Data::Root::RootDataIO RootIO("trtr");
-  RootIO.writeData(sample, "RootReaderTest-output.root");
+  // Create ROOT file with wrong branch names
+  const char *FileName = "WrongFormat.root";
+  const char *TreeName = "data";
+  TFile TheFile(FileName, "RECREATE");
+  TTree TheChain(TreeName, TreeName);
+  double SomeDouble;
+  TheChain.Branch("SomeDouble", &SomeDouble);
+  for (int i = 0; i < 5; i++)
+    TheChain.Fill();
+  TheChain.Write();
+  TheFile.Close();
 
-  auto sampleIn(RootIO.readData("RootReaderTest-output.root"));
-  BOOST_CHECK_EQUAL(sample.size(), sampleIn.size());
+  try {
+    ComPWA::Data::Root::readData(FileName, TreeName);
+  } catch (const std::runtime_error &e) {
+    BOOST_CHECK_EQUAL(e.what(), "Root::readData() | TTree does not have a "
```

Comment:
In that case it would be better to throw more precise exceptions and compare the exception type instead of the error message

Reply:
Added `CorruptFile` exception